### PR TITLE
ID-746 Let FastPass self-heal before firing event

### PR DESCRIPTION
**Pull request by tlangs**
## Summary of changes:
Only fire Propagation Delay event if `checkBucketReadAccess` fails twice or more.

### What
- Now that Rawls will issue a FastPass when a call to `checkBucketReadAccess` fails for many common reasons, we should give Terra a chance to self-heal permissions before firing off an event. If FastPass doesn't fix the issue, we should still get an event.

### Why
- Now that Terra self-heals, we should not be sending events for expected behavior.

### Testing strategy
- [ ]

### CODE REVIEW

File: src/pages/workspaces/workspace/WorkspaceContainer.js [hunk 1]
```diff
@@ -164,15 +164,19 @@ export const WorkspaceContainer = ({
...
     if (isGoogleWorkspaceSyncing) {
-      Ajax().Metrics.captureEvent(Events.permissionsSynchronizationDelayDisplayed, {
-        accessLevel: workspace.accessLevel,
-        createdDate: workspace.workspace.createdDate,
-        isWorkspaceCreator: workspace.workspace.createdBy === getUser().email,
-        ...extractWorkspaceDetails(workspace),
-      });
+      setWorkspaceSyncingChecks(workspaceSyncingChecks + 1);
+      if (workspaceSyncingChecks >= 2) {
+        Ajax().Metrics.captureEvent(Events.permissionsSynchronizationDelay, {
```
Comment (cahrens) [inline src/pages/workspaces/workspace/WorkspaceContainer.js]: I don't understand the point of doing this-- the event is just a metric about whether the user is setting the message that they don't have access. It has no interaction with FastPass itself.

Also, doing an increment sort of thing here doesn't make sense. This is an effect that is only called when the value of `isGoogleWorkspaceSyncing` changes. I believe the impact of this change is that the metric will never be emitted.

**Review by cahrens — CHANGES_REQUESTED** [review]

File: src/pages/workspaces/workspace/WorkspaceContainer.js (hunk 1 above)
Comment (cahrens) [reply src/pages/workspaces/workspace/WorkspaceContainer.js]: One possible solution would be to set a timeout for a given time interval, and then emit the event if `isGoogleWorkspaceSyncing` is still true when the timeout fires. But I think you might need to rethink this event in general-- what is it that you are trying to capture at this point? How many users see the message, or the length of time that they see the message?

Comment (tlangs) [reply src/pages/workspaces/workspace/WorkspaceContainer.js]: What's happening here is that UI fires off a `checkBucketReadAccess` request to Rawls, and if Rawls finds an issue with the access, it'll issue a FastPass grant. It'll still return an error to the UI, since permissions propagations aren't immediate. The UI then waits some amount of time, and calls `checkBucketReadAccess` again. Rawls does the checks, and if everything checks out, returns a success. We only want to alert if the second call to `checkBucketReadAccess` fails, so that the FastPass has time to take effect.

I could see Rawls returning a success if it vends a FastPass, but that doesn't _guarantee_ that the permissions are fixed. It can still take "sometimes 7 minutes or longer" for a IAM permissions to propagate.

Comment (tlangs) [reply src/pages/workspaces/workspace/WorkspaceContainer.js]: But, I do see that its looking at `isGoogleWorkspaceSyncing` changing, so yeah, its not having the intended effect!

File: src/libs/events.ts [hunk 2]
```diff
@@ -20,6 +20,7 @@ const eventsList = {
   analysisEnableBeta: 'analysis:enable',
   analysisDisableBeta: 'analysis:disable',
   analysisCreate: 'analysis:create',
+  analysisLaunch: 'analysis:launch',
```
Comment (cahrens) [inline src/libs/events.ts:23]: I also don't understand why these changes are in this PR.

Comment (tlangs) [reply src/libs/events.ts:23]: I was looking at events.js, decided to upgrade it to Typescript, and fixed the TS errors that cropped up. This value was referenced, but never actually defined in `events`.

**Comment by tlangs** [comment]
@cahrens I took a different approach and used a `store` instead. Does this seem like the right approach?

File: src/libs/events.ts (hunk 2 above)
Comment (cahrens) [reply src/libs/events.ts:23]: I don't know if we want to do conversions where the file extension is just changed without types added.

I also don't understand how `analysisLaunch` is being referenced if it isn't defined. Peculiar.

**Comment by tlangs** [comment]
Actually, wait. I can contain all this logic in `useWorkspace`...

File: src/libs/state.ts [hunk 3]
```diff
@@ -145,6 +145,8 @@ export const contactUsActive = atom(false);
 
 export const workspaceStore = atom<any>(undefined);
 
+export const workspaceCheckCloudAccessFailureStore = atom<number>(0);
```
Comment (cahrens) [inline src/libs/state.ts]: Doesn't this mean the same store will be used for all workspaces? 

I have to wonder again… is this change really necessary?

File: src/libs/events.ts (hunk 2 above)
Comment (tlangs) [reply src/libs/events.ts:23]: Because `events` wasn't typescript, there was no error when `https://github.com/DataBiosphere/terra-ui/blob/dev/src/analysis/AnalysisLauncher.js#L495` referenced it.

Comment (cahrens) [reply src/libs/events.ts:23]: Well actually, the place that calls the code isn't TypeScript so the fact that this file wasn't TypeScript doesn't fix the problem I think. But yes, the event should actually be defined (and/or IA should be given a heads-up)

File: src/libs/state.ts (hunk 3 above)
Comment (tlangs) [reply src/libs/state.ts]: Its not necessary, and been removed.

File: src/libs/events.ts (hunk 2 above)
Comment (tlangs) [reply src/libs/events.ts:23]: Gave IA a heads-up in slack!

File: src/pages/workspaces/workspace/useWorkspace.ts [hunk 4]
```diff
@@ -111,8 +114,16 @@ export const useWorkspace = (namespace, name): WorkspaceDetails => {
       } else {
         updateWorkspaceInStore(workspace, false);
         console.log('Google permissions are still syncing'); // eslint-disable-line no-console
+        if (times === 1) {
```
Comment (Ghost-in-a-Jar) [inline src/pages/workspaces/workspace/useWorkspace.ts:117]: Don't we want `>= 1`?

Comment (tlangs) [reply src/pages/workspaces/workspace/useWorkspace.ts:117]: We want the event to fire exactly once per workspace load. Basically, if FastPass can't help them within 15 seconds, we want to know about it. But we want to only know about it once.

**Review by Ghost-in-a-Jar — APPROVED** [review]

**Review by cahrens — APPROVED** [review]
Definitely a lot simpler! And thanks for the test.

File: src/libs/events.ts (hunk 2 above)
Comment (nawatts) [reply src/libs/events.ts:23]: Opened #4276 to add types to libs/events.
